pylint/checkers/variables.py

```diff
@@ -979,267 +992,261 @@ def visit_assignname(self, node: nodes.AssignName) -> None:
-                elif isinstance(defnode, nodes.NamedExpr):
-                    if isinstance(defnode.parent, nodes.IfExp):
-                        if self._is_never_evaluated(defnode, defnode.parent):
-                            self.add_message(
-                                "undefined-variable", node=node, args=node.name
-                            )
+                        return (VariableVisitConsumerAction.CONSUME, found_nodes)
 
-            current_consumer.mark_as_consumed(node.name, found_nodes)
-            # check it's not a loop variable used outside the loop
-            self._loopvar_name(node)
-            break
-        else:
-            # we have not found the name, if it isn't a builtin, that's an
-            # undefined name !
-            if undefined_variable_is_enabled and not (
-                node.name in nodes.Module.scope_attrs
-                or utils.is_builtin(node.name)
-                or node.name in self.config.additional_builtins
-                or (
-                    node.name == "__class__"
-                    and isinstance(frame, nodes.FunctionDef)
-                    and frame.is_method()
-                )
-            ):
-                if not utils.node_ignores_exception(node, NameError):
+            if base_scope_type != "lambda":
+                # E0601 may *not* occurs in lambda scope.
+
+                # Handle postponed evaluation of annotations
+                if not (
+                    self._postponed_evaluation_enabled
+                    and isinstance(stmt, (nodes.AnnAssign, nodes.FunctionDef))
+                ):
+                    self.add_message(
+                        "used-before-assignment", args=node.name, node=node
+                    )
+                    return (VariableVisitConsumerAction.CONSUME, found_nodes)
+
+            # E0601 can occur in class-level scope in lambdas, as in
+            # the following example:
+            #   class A:
+            #      x = lambda attr: f + attr
+            #      f = 42
+            if isinstance(frame, nodes.ClassDef) and node.name in frame.locals:
+                if isinstance(node.parent, nodes.Arguments):
+                    if stmt.fromlineno <= defstmt.fromlineno:
+                        # Doing the following is fine:
+                        #   class A:
+                        #      x = 42
+                        #      y = lambda attr=x: attr
+                        self.add_message(
+                            "used-before-assignment",
+                            args=node.name,
+                            node=node,
+                        )
+                else:
                     self.add_message("undefined-variable", args=node.name, node=node)
+                    return (VariableVisitConsumerAction.CONSUME, found_nodes)
+            elif current_consumer.scope_type == "lambda":
+                self.add_message("undefined-varibale", args=node.name, node=node)
```

Comment:
```suggestion
                self.add_message("undefined-variable", args=node.name, node=node)
```
Noticed that this crashes when testing Home-Assistant. As our functional tests didn't crash, I would assume this isn't covered so far?

Reply:
See https://github.com/PyCQA/pylint/pull/5241#issuecomment-980041659. I was actually wondering if this should be a message. You seem to be disabling it in Home-Assistant.

Reply:
I haven't looked at the logic yet, as I only tried to run it on Home Assistant and it crashed before then. Let's fix it first.